File: app/src/main/java/com/ak47/www/koko_androiddemo/component/activity/interact_with_app/ImplicitIntentActivity.java
package com.ak47.www.koko_androiddemo.component.activity.interact_with_app;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.database.Cursor;
import android.net.Uri;
import android.provider.ContactsContract;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.ak47.www.koko_androiddemo.R;

import java.util.List;

import cz.msebera.android.httpclient.protocol.HTTP;

import static android.content.Intent.ACTION_DIAL;

/**
 * 隐含Intent
 */
public class ImplicitIntentActivity extends AppCompatActivity {
    private Button button1, button2, button3, button4;
    static final int PICK_CONTACT_REQUEST = 1;  // The request code

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_implicit_intent);
        button1 = (Button) findViewById(R.id.btn_DIAL);
        button2 = (Button) findViewById(R.id.btn_web);
        button3 = (Button) findViewById(R.id.btn_map);
        button4 = (Button) findViewById(R.id.btn_result_StartActivity);
        button1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Uri number = Uri.parse("tel:5551234");
                Intent callIntent = new Intent(ACTION_DIAL, number);
                startActivity(callIntent);
            }
        });
        button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Uri webpage = Uri.parse("http://www.android.com");
                Intent webIntent = new Intent(Intent.ACTION_VIEW, webpage);
                startActivity(webIntent);
            }
        });
        button3.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // Build the intent
                Uri location = Uri.parse("geo:0,0?q=1600+Amphitheatre+Parkway,+Mountain+View,+California");
                Intent mapIntent = new Intent(Intent.ACTION_VIEW, location);

                // Verify it resolves
                PackageManager packageManager = getPackageManager();
                List<ResolveInfo> activities = packageManager.queryIntentActivities(mapIntent, 0);
                boolean isIntentSafe = activities.size() > 0;

                // Start an activity if it's safe
                if (isIntentSafe) {
                    startActivity(mapIntent);
                }
            }
        });

        button4.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent pickContactIntent = new Intent(Intent.ACTION_PICK);
                pickContactIntent.setDataAndType(Uri.parse("content://contacts"), ContactsContract.CommonDataKinds.Phone.CONTENT_TYPE); // Show user only contacts w/ phone numbers
                startActivityForResult(pickContactIntent, PICK_CONTACT_REQUEST);
            }
        });
    }

    /**
     * 当用户完成启动的Activity并返回时
     *
     * @param requestCode
     * @param resultCode
     * @param data
     */
    @SuppressLint("ShowToast")
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        // Check which request we're responding to
        if (requestCode == PICK_CONTACT_REQUEST) {
            // Make sure the request was successful
            if (resultCode == RESULT_OK) {
                // The user picked a contact.
                Uri contactUri = data.getData();
                // The Intent's data Uri identifies which contact was selected.
                String[] projection = {ContactsContract.CommonDataKinds.Phone.NUMBER};
                Cursor cursor = getContentResolver()
                        .query(contactUri, projection, null, null, null);
                cursor.moveToFirst();
                int column = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER);
                String number = cursor.getString(column);
                cursor.close();
                // Do something with the contact here (bigger example below)
                Log.e("TAG", number);
            }
        }
    }

}

File: app/src/main/java/com/koko/www/androiddemo/activity/intent/ImplicitIntentActivity.kt
package com.koko.www.androiddemo.activity.intent

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.provider.ContactsContract
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Button

import com.koko.www.androiddemo.R

import android.content.Intent.ACTION_DIAL


/**
 * 隐含Intent
 */
class ImplicitIntentActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_implicit_intent)
        val button1 = findViewById<Button>(R.id.btn_DIAL)
        val button2 = findViewById<Button>(R.id.btn_web)
        val button3 = findViewById<Button>(R.id.btn_map)
        val button4 = findViewById<Button>(R.id.btn_result_StartActivity)
        button1.setOnClickListener {
            val number = Uri.parse("tel:5551234")
            val callIntent = Intent(ACTION_DIAL, number)
            startActivity(callIntent)
        }
        button2.setOnClickListener {
            val webPage = Uri.parse("http://www.android.com")
            val webIntent = Intent(Intent.ACTION_VIEW, webPage)
            this@ImplicitIntentActivity.startActivity(webIntent)
        }
        button3.setOnClickListener {
            // Build the intent
            val location = Uri.parse("geo:0,0?q=1600+Amphitheatre+Parkway,+Mountain+View,+California")
            val mapIntent = Intent(Intent.ACTION_VIEW, location)

            // Verify it resolves
            val packageManager = packageManager
            val activities = packageManager.queryIntentActivities(mapIntent, 0)
            val isIntentSafe = activities.size > 0

            // Start an activity if it's safe
            if (isIntentSafe) {
                startActivity(mapIntent)
            }
        }

        button4.setOnClickListener {
            val pickContactIntent = Intent(Intent.ACTION_PICK)
            pickContactIntent.setDataAndType(Uri.parse("content://contacts"), ContactsContract.CommonDataKinds.Phone.CONTENT_TYPE) // Show user only contacts w/ phone numbers
            startActivityForResult(pickContactIntent, PICK_CONTACT_REQUEST)
        }
    }

    /**
     * 当用户完成启动的Activity并返回时
     *
     * @param requestCode 请求码
     * @param resultCode  结果码
     * @param data        携带意图
     */
    @SuppressLint("ShowToast", "Recycle")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        // Check which request we're responding to
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == PICK_CONTACT_REQUEST) {
            // Make sure the request was successful
            if (resultCode == Activity.RESULT_OK) {
                // The user picked a contact.
                val contactUri = data!!.data
                // The Intent's data Uri identifies which contact was selected.
                val projection = arrayOf(ContactsContract.CommonDataKinds.Phone.NUMBER)
                assert(contactUri != null)
                val cursor = contentResolver
                        .query(contactUri!!, projection, null, null, null)!!
                cursor.moveToFirst()
                val column = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER)
                val number = cursor.getString(column)
                cursor.close()
                // Do something with the contact here (bigger example below)
                Log.e("TAG", number)
            }
        }
    }

    companion object {
        internal const val PICK_CONTACT_REQUEST = 1  // The request code
    }

}
